lib/stdlib/src/erl_tar.erl

```diff
@@ -225,809 +280,1613 @@ month(10) -> "Oct";
-						% (or all).
-	 output = file   :: 'file' | 'memory',
-	 open_mode = [],			% Open mode options.
-	 verbose = false :: boolean()}).	% Verbose on/off.
+format_string(String, Size) when length(String) > Size ->
+    throw({error, {write_string, field_too_long}});
+format_string(String, Size) ->
+    Ascii = to_ascii(String),
+    if byte_size(Ascii) < Size ->
+            [Ascii, 0];
+       true ->
+            Ascii
+    end.
 
-extract_opts(List) ->
-    extract_opts(List, default_options()).
+format_octal(Octal) ->
+    iolist_to_binary(io_lib:fwrite("~.8B", [Octal])).
+
+add_header(#reader{}=Reader, #tar_header{}=Header, Opts) ->
+    {ok, Iodata} = build_header(Header, Opts),
+    do_write(Reader, Iodata).
+
+write_to_block(Block, IoData, Start) when is_list(IoData) ->
+    write_to_block(Block, iolist_to_binary(IoData), Start);
+write_to_block(Block, Bin, Start) when is_binary(Bin) ->
+    Size = byte_size(Bin),
+    <<Head:Start/unit:8, _:Size/unit:8, Rest/binary>> = Block,
+    <<Head:Start/unit:8, Bin/binary, Rest/binary>>.
+
+build_header(#tar_header{}=Header, Opts) ->
+    #tar_header{
+       name=Name,
+       mode=Mode,
+       uid=Uid,
+       gid=Gid,
+       size=Size,
+       typeflag=Type,
+       linkname=Linkname,
+       uname=Uname,
+       gname=Gname,
+       devmajor=Devmaj,
+       devminor=Devmin
+      } = Header,
+    Mtime = datetime_to_posix(Header#tar_header.mtime),
+    Xattrs = maps:to_list(Header#tar_header.xattrs),
+
+    Block0 = ?ZERO_BLOCK,
+    {Block1, Pax0} = write_string(Block0, ?V7_NAME, ?V7_NAME_LEN, Name, ?PAX_PATH, #{}),
+    Block2 = write_octal(Block1, ?V7_MODE, ?V7_MODE_LEN, Mode),
+    {Block3, Pax1} = write_numeric(Block2, ?V7_UID, ?V7_UID_LEN, Uid, ?PAX_UID, Pax0),
+    {Block4, Pax2} = write_numeric(Block3, ?V7_GID, ?V7_GID_LEN, Gid, ?PAX_GID, Pax1),
+    {Block5, Pax3} = write_numeric(Block4, ?V7_SIZE, ?V7_SIZE_LEN, Size, ?PAX_SIZE, Pax2),
+    {Block6, Pax4} = write_numeric(Block5, ?V7_MTIME, ?V7_MTIME_LEN, Mtime, ?PAX_NONE, Pax3),
+    {Block7, Pax5} = write_string(Block6, ?V7_TYPE, ?V7_TYPE_LEN, <<Type>>, ?PAX_NONE, Pax4),
+    {Block8, Pax6} = write_string(Block7, ?V7_LINKNAME, ?V7_LINKNAME_LEN,
+                                  Linkname, ?PAX_LINKPATH, Pax5),
+    {Block9, Pax7} = write_string(Block8, ?USTAR_UNAME, ?USTAR_UNAME_LEN,
+                                  Uname, ?PAX_UNAME, Pax6),
+    {Block10, Pax8} = write_string(Block9, ?USTAR_GNAME, ?USTAR_GNAME_LEN,
+                                   Gname, ?PAX_GNAME, Pax7),
+    {Block11, Pax9} = write_numeric(Block10, ?USTAR_DEVMAJ, ?USTAR_DEVMAJ_LEN,
+                                    Devmaj, ?PAX_NONE, Pax8),
+    {Block12, Pax10} = write_numeric(Block11, ?USTAR_DEVMIN, ?USTAR_DEVMIN_LEN,
+                                     Devmin, ?PAX_NONE, Pax9),
+    {Block13, Pax11} = set_path(Block12, Pax10),
+    Pax12 = lists:foldl(fun ({K,V}, Acc) ->
+                                maps:put(<<?PAX_XATTR_STR,K>>, V, Acc)
```

Comment:
Great catch! As it turns out, we'll never be writing any xattrs which are not PAX-standard, so this code can be removed anyway.